engine/classes/Elgg/Database/EntityTable.php

```diff
@@ -217,13 +296,32 @@ public function get($guid, $type = '') {
 			return false;
 		}
 
-		// Check local cache first
-		$new_entity = _elgg_services()->entityCache->get($guid);
-		if ($new_entity) {
+		$guid = (int) $guid;
+
+		$memcache = _elgg_get_memcache('new_entity_cache');
+		
+		$entity = $this->entity_cache->get($guid);
+		if (!$entity) {
+			$entity = $memcache->load($guid);
+		}
+
+		if (!$entity instanceof \ElggEntity) {
+			$entity = false;
+		}
+
+		// Validate accessibility
+		if ($entity && !elgg_get_ignore_access() && !has_access_to_entity($entity)) {
```

Comment:
> For in-memory we were refusing to cache when access was off and clearing cache on session change. That logic may have got broken along the way.

That was true for caching, but false for session change invalidation. I have made the necessary changes and added tests. On their way.